File: apps/desktop/src/stores/repositoryGroups.ts
import { create } from "zustand";
import type { RepositoryGroup } from "../types";
import * as api from "../api";

interface RepositoryGroupsState {
  // State
  groups: RepositoryGroup[];
  selectedGroupId: string | null;
  isLoading: boolean;
  hasFetched: boolean;
  error: string | null;

  // Computed
  selectedGroup: () => RepositoryGroup | undefined;

  // Actions
  fetchGroups: (workspaceId?: string) => Promise<void>;
  createGroup: (workspaceId: string, name?: string, repositoryIds?: string[]) => Promise<RepositoryGroup>;
  updateGroup: (id: string, name?: string) => Promise<RepositoryGroup>;
  deleteGroup: (id: string) => Promise<void>;
  selectGroup: (id: string | null) => void;
  addRepositoryToGroup: (groupId: string, repositoryId: string) => Promise<void>;
  removeRepositoryFromGroup: (groupId: string, repositoryId: string) => Promise<void>;
  getOrCreateSingleRepoGroup: (workspaceId: string, repositoryId: string) => Promise<string>;
  clearError: () => void;
}

export const useRepositoryGroupsStore = create<RepositoryGroupsState>((set, get) => ({
  // Initial state
  groups: [],
  selectedGroupId: null,
  isLoading: false,
  hasFetched: false,
  error: null,

  // Computed
  selectedGroup: () => {
    const { groups, selectedGroupId } = get();
    return groups.find((g) => g.id === selectedGroupId);
  },

  // Actions
  fetchGroups: async (workspaceId?: string) => {
    try {
      set({ isLoading: true, error: null });
      const groups = await api.listRepositoryGroups(workspaceId);
      set({ groups, isLoading: false, hasFetched: true });
    } catch (error) {
      set({
        error:
          error instanceof Error ? error.message : "Failed to fetch repository groups",
        isLoading: false,
        hasFetched: true,
      });
    }
  },

  createGroup: async (workspaceId: string, name?: string, repositoryIds?: string[]) => {
    try {
      set({ isLoading: true, error: null });
      const group = await api.createRepositoryGroup(workspaceId, name, repositoryIds);
      set((state) => ({
        groups: [...state.groups, group],
        isLoading: false,
      }));
      return group;
    } catch (error) {
      const message =
        error instanceof Error ? error.message : "Failed to create repository group";
      set({ error: message, isLoading: false });
      throw new Error(message);
    }
  },

  updateGroup: async (id: string, name?: string) => {
    try {
      set({ isLoading: true, error: null });
      const group = await api.updateRepositoryGroup(id, name);
      set((state) => ({
        groups: state.groups.map((g) =>
          g.id === id ? group : g
        ),
        isLoading: false,
      }));
      return group;
    } catch (error) {
      const message =
        error instanceof Error ? error.message : "Failed to update repository group";
      set({ error: message, isLoading: false });
      throw new Error(message);
    }
  },

  deleteGroup: async (id: string) => {
    try {
      set({ isLoading: true, error: null });
      await api.deleteRepositoryGroup(id);
      set((state) => ({
        groups: state.groups.filter((g) => g.id !== id),
        selectedGroupId:
          state.selectedGroupId === id ? null : state.selectedGroupId,
        isLoading: false,
      }));
    } catch (error) {
      set({
        error:
          error instanceof Error ? error.message : "Failed to delete repository group",
        isLoading: false,
      });
    }
  },

  selectGroup: (id: string | null) => {
    set({ selectedGroupId: id });
  },

  addRepositoryToGroup: async (groupId: string, repositoryId: string) => {
    try {
      set({ isLoading: true, error: null });
      await api.addRepositoryToGroup(groupId, repositoryId);
      set((state) => ({
        groups: state.groups.map((g) =>
          g.id === groupId
            ? { ...g, repository_ids: [...g.repository_ids, repositoryId] }
            : g
        ),
        isLoading: false,
      }));
    } catch (error) {
      set({
        error:
          error instanceof Error ? error.message : "Failed to add repository to group",
        isLoading: false,
      });
    }
  },

  removeRepositoryFromGroup: async (groupId: string, repositoryId: string) => {
    try {
      set({ isLoading: true, error: null });
      await api.removeRepositoryFromGroup(groupId, repositoryId);
      set((state) => ({
        groups: state.groups.map((g) =>
          g.id === groupId
            ? { ...g, repository_ids: g.repository_ids.filter((id) => id !== repositoryId) }
            : g
        ),
        isLoading: false,
      }));
    } catch (error) {
      set({
        error:
          error instanceof Error ? error.message : "Failed to remove repository from group",
        isLoading: false,
      });
    }
  },

  getOrCreateSingleRepoGroup: async (workspaceId: string, repositoryId: string) => {
    try {
      set({ isLoading: true, error: null });
      const groupId = await api.getOrCreateSingleRepoGroup(workspaceId, repositoryId);

      // Refresh groups to include the new group
      const groups = await api.listRepositoryGroups(workspaceId);
      set({ groups, isLoading: false });

      return groupId;
    } catch (error) {
      const message =
        error instanceof Error ? error.message : "Failed to get or create single repo group";
      set({ error: message, isLoading: false });
      throw new Error(message);
    }
  },

  clearError: () => set({ error: null }),
}));

File: apps/desktop/src-tauri/src/commands/repository_groups.rs
use std::sync::Arc;

use tauri::State;

use crate::{entities::RepositoryGroup, services::AppState};

/// Lists all repository groups
#[tauri::command]
pub async fn list_repository_groups(
    state: State<'_, Arc<AppState>>,
    workspace_id: Option<String>,
) -> Result<Vec<RepositoryGroup>, String> {
    match workspace_id {
        Some(id) => state
            .repository_group_service
            .list_by_workspace(&id)
            .await
            .map_err(|e| e.to_string()),
        None => state
            .repository_group_service
            .list()
            .await
            .map_err(|e| e.to_string()),
    }
}

/// Gets a repository group by ID
#[tauri::command]
pub async fn get_repository_group(
    state: State<'_, Arc<AppState>>,
    id: String,
) -> Result<Option<RepositoryGroup>, String> {
    state
        .repository_group_service
        .get(&id)
        .await
        .map_err(|e| e.to_string())
}

/// Creates a new repository group
#[tauri::command]
pub async fn create_repository_group(
    state: State<'_, Arc<AppState>>,
    workspace_id: String,
    name: Option<String>,
    repository_ids: Vec<String>,
) -> Result<RepositoryGroup, String> {
    let group_id = uuid::Uuid::new_v4().to_string();
    let mut group = RepositoryGroup::new(group_id.clone(), workspace_id);

    if let Some(n) = name {
        group = group.with_name(n);
    }

    group.repository_ids = repository_ids;

    state
        .repository_group_service
        .create(&group)
        .await
        .map_err(|e| e.to_string())?;

    Ok(group)
}

/// Updates a repository group
#[tauri::command]
pub async fn update_repository_group(
    state: State<'_, Arc<AppState>>,
    id: String,
    name: Option<String>,
) -> Result<RepositoryGroup, String> {
    let mut group = state
        .repository_group_service
        .get(&id)
        .await
        .map_err(|e| e.to_string())?
        .ok_or("Repository group not found")?;

    group.name = name;
    group.updated_at = chrono::Utc::now();

    state
        .repository_group_service
        .update(&group)
        .await
        .map_err(|e| e.to_string())?;

    Ok(group)
}

/// Deletes a repository group
#[tauri::command]
pub async fn delete_repository_group(
    state: State<'_, Arc<AppState>>,
    id: String,
) -> Result<(), String> {
    state
        .repository_group_service
        .delete(&id)
        .await
        .map_err(|e| e.to_string())
}

/// Adds a repository to a repository group
#[tauri::command]
pub async fn add_repository_to_group(
    state: State<'_, Arc<AppState>>,
    group_id: String,
    repository_id: String,
) -> Result<(), String> {
    state
        .repository_group_service
        .add_repository(&group_id, &repository_id)
        .await
        .map_err(|e| e.to_string())
}

/// Removes a repository from a repository group
#[tauri::command]
pub async fn remove_repository_from_group(
    state: State<'_, Arc<AppState>>,
    group_id: String,
    repository_id: String,
) -> Result<(), String> {
    state
        .repository_group_service
        .remove_repository(&group_id, &repository_id)
        .await
        .map_err(|e| e.to_string())
}

/// Gets or creates a single-repo group for a repository
/// This is used when creating tasks for a single repository
#[tauri::command]
pub async fn get_or_create_single_repo_group(
    state: State<'_, Arc<AppState>>,
    workspace_id: String,
    repository_id: String,
) -> Result<String, String> {
    state
        .repository_group_service
        .get_or_create_single_repo_group(&workspace_id, &repository_id)
        .await
        .map_err(|e| e.to_string())
}

/// Lists files in all repositories of a repository group for autocomplete
/// Uses git ls-files to get tracked files, respecting .gitignore
#[tauri::command]
pub async fn list_repository_group_files(
    state: State<'_, Arc<AppState>>,
    repository_group_id: String,
    query: Option<String>,
    limit: Option<usize>,
) -> Result<Vec<String>, String> {
    use std::path::PathBuf;

    // Get the repository group
    let group = state
        .repository_group_service
        .get(&repository_group_id)
        .await
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "Repository group not found".to_string())?;

    let mut all_files: Vec<String> = Vec::new();

    // Collect files from all repositories in the group
    for repo_id in &group.repository_ids {
        let repo = match state
            .repository_service
            .get(repo_id)
            .await
            .map_err(|e| e.to_string())?
        {
            Some(r) => r,
            None => continue, // Skip if repository not found
        };

        let repo_path = PathBuf::from(&repo.local_path);
        if !repo_path.exists() {
            continue; // Skip if repository path doesn't exist
        }

        // Use git ls-files to get tracked files
        let output = match tokio::process::Command::new("git")
            .args(["ls-files"])
            .current_dir(&repo_path)
            .output()
            .await
        {
            Ok(o) => o,
            Err(_) => continue, // Skip on error
        };

        if !output.status.success() {
            continue; // Skip on git error
        }

        let files_str = String::from_utf8_lossy(&output.stdout);
        // Prefix files with repository name for multi-repo groups
        let repo_prefix = if group.repository_ids.len() > 1 {
            format!("{}:", repo.name)
        } else {
            String::new()
        };

        for line in files_str.lines() {
            all_files.push(format!("{}{}", repo_prefix, line));
        }
    }

    // Filter by query if provided (fuzzy match)
    if let Some(q) = query {
        let q_lower = q.to_lowercase();
        all_files.retain(|f| {
            let f_lower = f.to_lowercase();
            // Match if query appears anywhere in the path
            // or if the file name contains the query
            f_lower.contains(&q_lower)
                || f.split('/')
                    .next_back()
                    .map(|name| name.to_lowercase().contains(&q_lower))
                    .unwrap_or(false)
        });

        // Sort by relevance: exact filename matches first, then by path length
        all_files.sort_by(|a, b| {
            let a_name = a.split('/').next_back().unwrap_or(a).to_lowercase();
            let b_name = b.split('/').next_back().unwrap_or(b).to_lowercase();
            let a_exact = a_name == q_lower;
            let b_exact = b_name == q_lower;
            let a_starts = a_name.starts_with(&q_lower);
            let b_starts = b_name.starts_with(&q_lower);

            match (a_exact, b_exact) {
                (true, false) => std::cmp::Ordering::Less,
                (false, true) => std::cmp::Ordering::Greater,
                _ => match (a_starts, b_starts) {
                    (true, false) => std::cmp::Ordering::Less,
                    (false, true) => std::cmp::Ordering::Greater,
                    _ => a.len().cmp(&b.len()),
                },
            }
        });
    }

    // Apply limit
    let limit = limit.unwrap_or(50);
    all_files.truncate(limit);

    Ok(all_files)
}
